pkg/controllermanager/controller/shoot/shoot_reference_control.go

```diff
@@ -0,0 +1,213 @@
+// deletions as long as they are still referenced.
+func (r *shootReferenceReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	ctx := context.TODO()
+
+	gardenClient, err := r.clientMap.GetClient(ctx, keys.ForGarden())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	shoot := &gardencorev1beta1.Shoot{}
+	if err := gardenClient.Client().Get(ctx, request.NamespacedName, shoot); err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			return reconcile.Result{}, err
+		}
+		r.logger.Debugf("shoot %q is not found, not trying to reconcile", request.Name)
+		return reconcile.Result{}, nil
+	}
+
+	r.logger.Infof("[SHOOT REFERENCE CONTROL] %s", request)
+
+	if err := r.reconcileShootReferences(ctx, shoot); err != nil {
+		return reconcile.Result{}, err
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *shootReferenceReconciler) reconcileShootReferences(ctx context.Context, shoot *gardencorev1beta1.Shoot) error {
+	gardenClient, err := r.clientMap.GetClient(ctx, keys.ForGarden())
+	if err != nil {
+		return fmt.Errorf("failed to get garden client: %w", err)
+	}
+
+	if err := releaseUnreferencedSecrets(ctx, gardenClient.Client(), shoot); err != nil {
+		return err
+	}
+
+	if shoot.DeletionTimestamp != nil {
+		return controllerutils.RemoveFinalizer(ctx, gardenClient.Client(), shoot, FinalizerName)
+	}
+
+	var (
+		shootNeedsFinalizer bool
+		dnsSecretNames      = secretNamesForDNSProviders(shoot)
+	)
+
+	for _, dnsSecretName := range dnsSecretNames {
+		shootNeedsFinalizer = true
+		secret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: shoot.Namespace,
+				Name:      dnsSecretName,
+			},
+		}
+		if err := controllerutils.EnsureFinalizer(ctx, gardenClient.Client(), secret, FinalizerName); client.IgnoreNotFound(err) != nil {
+			return err
+		}
+	}
+
+	if shootNeedsFinalizer {
+		return controllerutils.EnsureFinalizer(ctx, gardenClient.Client(), shoot, FinalizerName)
+	}
+
+	return controllerutils.RemoveFinalizer(ctx, gardenClient.Client(), shoot, FinalizerName)
+}
+
+func releaseUnreferencedSecrets(ctx context.Context, gardenClient client.Client, shoot *gardencorev1beta1.Shoot) error {
+	secretNames, err := getUnreferencedSecretNames(ctx, gardenClient, shoot)
+	if err != nil {
+		return err
+	}
+
+	for _, secretName := range secretNames {
+		secret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      secretName,
+				Namespace: shoot.Namespace,
+			},
+		}
+		if err := controllerutils.RemoveFinalizer(ctx, gardenClient, secret, FinalizerName); client.IgnoreNotFound(err) != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func getUnreferencedSecretNames(ctx context.Context, gardenClient client.Client, shoot *gardencorev1beta1.Shoot) ([]string, error) {
+	namespace := shoot.Namespace
+
+	secrets := &corev1.SecretList{}
+	if err := gardenClient.List(ctx, secrets, client.InNamespace(namespace)); err != nil {
+		return nil, err
+	}
```

Comment:
As discussed on the phone, I am slightly worried about the performance aspect of looping over all Secrets in the Shoot namespace for every update of a given Shoot.
So let's
- use two separate queues/reconcilers for Shoots and Secrets to not fan-out compute effort on each Shoot update and rather enqueue referenced secrets
- only enqueue Shoots, if the relevant spec section changed (for now `spec.dns.providers`)

Reply:
I had a second thought about this and think the solution with a separate queue/reconciler is still elegant but also intricate since the "protection mechanism" is distributed. When the GCM restarts we also have to iterate over every secret in the cluster of all namespaces which may be a large amount, too.

What we can alternately do, is to stick to the logic of this PR but instead of just adding a finalizer to secrets we also label them at the same time. We can then make the reconciler request only the labelled secrets in the project namespace, instead of requesting all. WDYT?

Reply:
That's a valid concern, got it.
What about only enqueuing Secrets if they have a deletion timestamp (for both add and update)? Then we wouldn't need to iterate over all secrets on startup and only act on the ones we care about (deleting secrets without finalizers will only cause a delete event and no update event, IIRC).
We will need to iterate over all Shoots on startup anyways and we can then enqueue the referenced Secrets into the Secret queue and the secret reconciler will then take over setting/removing the finalizer. WDYT?

Reply:
Any ad-hoc concerns about the labelling approach?

Reply:
One concern would be, that users might remove the label (e.g. by using `k replace` instead of `k apply`).
If we rely on that label being present for releasing the secret, we might miss some secrets we should release.